atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/LightweightOppToken.java

```diff
@@ -25,15 +26,35 @@
 import com.palantir.logsafe.SafeArg;
 import java.nio.ByteBuffer;
 import java.util.Arrays;
+import java.util.Locale;
 
 public class LightweightOppToken implements Comparable<LightweightOppToken> {
+    private static final BaseEncoding lowerHex = BaseEncoding.base16().lowerCase();
 
     final byte[] bytes;
 
     public LightweightOppToken(byte[] bytes) {
         this.bytes = bytes;
     }
 
+    public static LightweightOppToken fromHex(String token) {
+        String normalized = hasUppercase(token)
+                ? token.toLowerCase(Locale.ROOT) // OPP tokens should be lowercase already, normalize if needed
```

Comment:
I don't believe `hasUppercase` is necessary due to [StringLatin1.java#L430-L431](https://github.com/openjdk/jdk/blob/2116928528c0554b2ba0171bd7968ab693972804/src/java.base/share/classes/java/lang/StringLatin1.java#L430-L431) and [StringUTF16.java#L811-L812](https://github.com/openjdk/jdk/blob/2116928528c0554b2ba0171bd7968ab693972804/src/java.base/share/classes/java/lang/StringUTF16.java#L811-L812) where the String implementation avoids allocation if the input is returned.

I think the important part is matching the expected inputs to avoid creating new strings as often as possible.

Perhaps we could check if there are any upper-case characters, and used a `BaseEncoding.base16()` instance, otherwise use `BaseEncoding.base16().lowerCase()`, avoiding string re-allocation.

Reply:
I like that, we can pick decoder based on presence of uppercase (though I think for cassandra these strings will always be lower case hex encoded already.

Reply:
Updated including fallback that converts to uppercase if the input was mixed case (mostly just to satisfy tests)